apps/settings/js/panels/simpin_dialog/simpin_dialog.js

```diff
@@ -0,0 +1,590 @@
+      lockType = lockType || 'pin';
+      var pin = this._elements.pinInput.value;
+      var newPin = this._elements.newPinInput.value;
+      var confirmPin = this._elements.confirmPinInput.value;
+
+      if (pin === '' || newPin === '' || confirmPin === '') {
+        return Promise.reject();
+      }
+      
+      if (newPin !== confirmPin) {
+        this._showMessage('newPinErrorMsg');
+        this._elements.newPinInput.value = '';
+        this._elements.confirmPinInput.value = '';
+        return Promise.reject();
+      }
+
+      return this._setCardLock({
+        lockType: lockType,
+        pin: pin,
+        newPin: newPin
+      });
+    },
+
+    /**
+     * This is an internal function to pass all arguments to icc and waiting
+     * for returned results.
+     *
+     * @memberOf SimPinDialog
+     * @param {Object} options
+     * @access private
+     * @return {Promise}
+     */
+    _setCardLock: function(options) {
+      var defer = Defer();
+      var req = this._icc.setCardLock(options);
+
+      req.onsuccess = () => {
+        defer.resolve();
+      };
+
+      req.onerror = () => {
+        this._handleCardLockError({
+          lockType: options.lockType,
+          retryCount: req.error.retryCount,
+          defer: defer
+        });
+      };
+
+      return defer.promise;
+    },
+
+    /**
+     * this function is used when we are going to update fdn contacts under
+     * call panel.
+     *
+     * @memberOf SimPinDialog
+     * @access private
+     * @return {Promise}
+     */
+    _updateFdnContact: function() {
+      //  Updates a FDN contact. For some reason, `icc.updateContact` requires
+      //  the  pin input value instead of delegating to `icc.setCardLock`.
+      //  That means that, in case of failure, the error is different that the
+      //  one that `icc.setCardLock` gives. This means that we have to handle
+      //  it separatedly instead of being able to use the existing
+      //  `handleCardLockError` above.
+      //
+      //  Among other things, it doesn't include the retryCount, so we can't
+      //  tell the user how many remaining tries she has. What a mess.
+      //  This should be solved when bug 1070941 is fixed.
+      var promise = new Promise((resolve, reject) => {
+        var fdnContactInfo = this._pinOptions.fdnContactInfo;
+        var req = this._icc.updateContact('fdn',
+          fdnContactInfo, this._elements.pinInput.value);
+
+        req.onsuccess = () => {
+          resolve(fdnContactInfo);
+        };
+
+        req.onerror = (e) => {
+          switch (req.error.name) {
+            case 'IncorrectPassword':
+              // TODO: count retries (not supported by the platform) ->
+              // Bug 1070941
+              this._initUI('get_pin2');
+              this._showMessage('fdnErrorMsg');
+              this._elements.pinInput.value = '';
+              this._elements.pinInput.focus();
+              reject();
```

Comment:
Calling to `reject` makes the dialog close. However, what we want here is to allow the user entering the pin code again, so we can't reject nor resolve here but only display the error message.